pkg/controller/daemon/util/daemonset_util.go

```diff
@@ -199,3 +199,51 @@ func ReplaceDaemonSetPodHostnameNodeAffinity(affinity *v1.Affinity, nodename str
 
 	return affinity
 }
+
+// AppendNoScheduleTolerationIfNotExist appends unscheduable toleration to `.spec` if not exist; otherwise,
+// no changes to `.spec.tolerations`.
+func AppendNoScheduleTolerationIfNotExist(tolerations []v1.Toleration) []v1.Toleration {
+	unschedulableToleration := v1.Toleration{
+		Key:      algorithm.TaintNodeUnschedulable,
+		Operator: v1.TolerationOpExists,
+		Effect:   v1.TaintEffectNoSchedule,
+	}
+
+	unschedulableTaintExist := false
+
+	for _, t := range tolerations {
+		if TolerationEqual(t, unschedulableToleration) {
+			unschedulableTaintExist = true
+			break
+		}
+	}
+
+	if !unschedulableTaintExist {
+		tolerations = append(tolerations, unschedulableToleration)
+	}
+
+	return tolerations
+}
+
+// TolerationEqual checks if two provided tolerations are equal or not.
+func TolerationEqual(first, second v1.Toleration) bool {
```

Comment:
This is mostly a dup [util/tolerations.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/util/tolerations/tolerations.go#L105)

Reply:
yes. I copied from there; it operates `api.*` instead of `v1.*`.